modules/platforms/cpp/core/src/impl/cluster/cluster_group_impl.cpp

```diff
@@ -95,6 +170,48 @@ namespace ignite
 
                 return SP_ComputeImpl(new compute::ComputeImpl(GetEnvironmentPointer(), computeProc));
             }
+
+            std::vector<ClusterNode> ClusterGroupImpl::RefreshNodes()
+            {
+                SharedPointer<interop::InteropMemory> memIn = GetEnvironment().AllocateMemory();
+                SharedPointer<interop::InteropMemory> memOut = GetEnvironment().AllocateMemory();
+                interop::InteropOutputStream out(memIn.Get());
+                binary::BinaryWriterImpl writer(&out, GetEnvironment().GetTypeManager());
+
+                CsLockGuard mtx(nodesLock);
```

Comment:
Locking scope is too large here. We only need to hold lock when updating local values - `topVer` and `nodes`.